src/init.jl

```diff
@@ -48,4 +48,84 @@ function __init__()
         export DiffEqIntegrator
 
     end
+
+    @require ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210" begin
+
+        import .ForwardDiff
+        DiffResults = ForwardDiff.DiffResults
+
+        function ∂ℓπ∂θ_forwarddiff(ℓπ, θ::AbstractVector)
+            res = DiffResults.GradientResult(θ)
+            ForwardDiff.gradient!(res, ℓπ, θ)
+            return DiffResults.value(res), DiffResults.gradient(res)
+        end
+
+        function ∂ℓπ∂θ_forwarddiff(ℓπ, θ::AbstractMatrix)
+            jacob = similar(θ)
+            res = DiffResults.JacobianResult(similar(θ, size(θ, 2)), jacob)
+            ForwardDiff.jacobian!(res, ℓπ, θ)
+            jacob_full = DiffResults.jacobian(res)
+            
+            d, n = size(jacob)
+            for i in 1:n
+                jacob[:,i] = jacob_full[i,1+(i-1)*d:i*d]
+            end
+            return DiffResults.value(res), jacob
+        end
+
+        # Implementation 2
+        # function ∂ℓπ∂θ_forwarddiff(ℓπ, θ::AbstractMatrix)
+        #     local densities
+        #     f(x) = (densities = ℓπ(x); sum(densities))
+        #     res = DiffResults.GradientResult(θ)
+        #     ForwardDiff.gradient!(res, f, θ)
+        #     return ForwardDiff.value.(densities), DiffResults.gradient(res)
+        # end
+
+        # Implementation 3
+        # function ∂ℓπ∂θ_forwarddiff(ℓπ, θ::AbstractMatrix)
+        #     v = similar(θ, size(θ, 2))
+        #     g = similar(θ)
+        #     for i in 1:size(θ, 2)
+        #         res = GradientResult(θ[:,i])
+        #         gradient!(res, ℓπ, θ[:,i])
+        #         v[i] = value(res)
+        #         g[:,i] = gradient(res)
+        #     end
+        #     return v, g
+        # end
+
+        function Hamiltonian(metric, ℓπ)
+            ∂ℓπ∂θ(θ::AbstractVecOrMat) = ∂ℓπ∂θ_forwarddiff(ℓπ, θ)
+            return Hamiltonian(metric, ℓπ, ∂ℓπ∂θ)
+        end
+
+    end
+
+    @require Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f" begin
+
+        import .Zygote
+
+        function ∂ℓπ∂θ_zygote(ℓπ, θ::AbstractVector)
+            res, back = Zygote.pullback(ℓπ, θ)
+            return res[1], back(1)[1]
+        end
+
+        function ∂ℓπ∂θ_zygote(ℓπ, θ::AbstractMatrix)
+            res, back = Zygote.pullback(ℓπ, θ)
+            return res, back(fill(1, size(θ)...))[1]
+        end
+
+        # NOTE
+        # It's important for Zygote to come after ForwardDiff otherwise
+        # the function def below will be override by ForwardDiff's def.
+        # This is probably because Zygote is dependent on ForwardDiff and
+        # they will be also available at the same time.
+        function Hamiltonian(metric, ℓπ)

```

Comment:
That seems wrong since it will override ForwardDiff's definition - so if both Zygote and ForwardDiff are loaded, it's never possible to actually use the constructor for ForwardDiff?